services/worker-manager/src/providers/azure.js

```diff
@@ -0,0 +1,456 @@
+        workerGroup: this.providerId,
+        workerId: virtualMachine.vmId,
+        created: now,
+        lastModified: now,
+        lastChecked: now,
+        expires: taskcluster.fromNow('1 week'),
+        state: this.Worker.states.REQUESTED,
+        capacity: cfg.capacityPerInstance,
+        providerData: {
+          ...providerData,
+          registrationExpiry,
+        },
+      });
+    }));
+  }
+
+  async deprovision({workerPool}) {
+    // nothing to do: we just wait for workers to terminate themselves
+  }
+
+  async registerWorker({worker, workerPool, workerIdentityProof}) {
+    const {document} = workerIdentityProof;
+
+    // use the same message for all errors here, so as not to give an attacker
+    // extra information.
+    const error = () => new ApiError('Signature validation error');
+
+    // workerIdentityProof is a signed message
+
+    // We need to check that:
+    // 1. The embedded document was signed with the private key corresponding to the
+    //    embedded public key
+    // 2. The embedded public key has a proper certificate chain back to a trusted CA
+    // 3. The embedded message contains the vmId that matches the worker making the request
+
+    // signature is base64-encoded DER-format PKCS#7 / CMS message
+
+    // decode base64, load DER, extract PKCS#7 message
+    let decodedMessage = Buffer.from(document, 'base64');
+    let message;
+    try {
+      let asn1 = forge.asn1.fromDer(forge.util.createBuffer(decodedMessage));
+      message = forge.pkcs7.messageFromAsn1(asn1);
+    } catch (err) {
+      this.monitor.warning('Error extracting PKCS#7 message', {error: err.toString()});
+      throw error();
+    }
+
+    let content, crt, pem, sig;
+    // get message content, signing certificate, and signature
+    try {
+      // in testing, message.content is empty, so we access the raw ASN1 structure
+      content = message.rawCapture.content.value[0].value;
+      // convert to pem for convenience
+      // there may be more than one certificate, we only look at the first one
+      crt = message.certificates[0];
+      pem = forge.pki.publicKeyToPem(crt.publicKey);
+      sig = message.rawCapture.signature;
+    } catch (err) {
+      this.monitor.warning('Error extracting PKCS#7 message content', {error: err.toString()});
+      throw error();
+    }
+
+    // verify that the message is properly signed
+    try {
+      let verifier = crypto.createVerify('RSA-SHA256');
+      verifier.update(Buffer.from(content));
+      assert(verifier.verify(pem, sig, 'binary'), true);
+    } catch (err) {
+      this.monitor.warning('Error verifying PKCS#7 message signature');
+      throw error();
+    }
+
+    // verify that the embedded certificate has proper chain of trust
+    try {
+      assert(forge.pki.verifyCertificateChain(this.caStore, [crt]), true);
+    } catch (err) {
+      this.monitor.warning('Error verifying certificate chain', {error: err.toString()});
```

Comment:
I'm not sure we need to do all of this `monitor.warning` stuff since we throw the error anyway right away and it will have a reference to where the error happened.

Reply:
I was wrong about this